src/misc/2fa/webauthn/WebauthnClient.ts

```diff
@@ -1,167 +1,149 @@
+import type {U2fChallenge} from "../../../api/entities/sys/U2fChallenge.js"
+import type {WebauthnResponseData} from "../../../api/entities/sys/WebauthnResponseData.js"
+import {createWebauthnResponseData} from "../../../api/entities/sys/WebauthnResponseData.js"
+import type {IWebauthn} from "./IWebauthn.js"
+import {U2F_APPID, U2f_APPID_SUFFIX, WEBAUTHN_RP_ID} from "./IWebauthn.js"
+import {U2fKey} from "../../../api/entities/sys/U2fKey.js"
 
-		if (window.location.hostname.endsWith("tutanota.com")) {
-			this.appId = "https://tutanota.com/u2f-appid.json"
-		} else {
-			this.appId = getHttpOrigin() + "/u2f-appid.json"
-		}
+/** Web authentication entry point for the rest of the app. */
+export interface IWebauthnClient {
+	isSupported(): Promise<boolean>;
+
+	/** Whether it's possible to attempt a challenge. It might not be possible if there are not keys for this domain. */
+	canAttemptChallenge(challenge: U2fChallenge): Promise<{canAttempt: Array<U2fKey>, cannotAttempt: Array<U2fKey>}>;
+
+	register(userId: Id, name: string, mailAddress: string): Promise<U2fRegisteredDevice>;
+
+	authenticate(challenge: U2fChallenge): Promise<WebauthnResponseData>;
+
+	abortCurrentOperation(): Promise<void>;
+}
+
+export class WebauthnClient implements IWebauthnClient {
+	constructor(
+		private readonly webauthn: IWebauthn,
+		private readonly clientWebRoot: string,
+	) {
 	}
 
-	isSupported(): boolean {
-		// We explicitly disable apps and desktop because even if webauthn somehow works there it won't match our domain.
-		// For those platforms another implementation with separate webview should be used instead.
-		return !isDesktop() && !isApp() &&
-			this.api != null &&
-			// @ts-ignore see polyfill.js
-			// We just stub BigInt in order to import cborg without issues but we can't actually use it
-			!BigInt.polyfilled
+	isSupported(): Promise<boolean> {
+		return this.webauthn.isSupported()
 	}
 
-	async register(userId: Id, name: string, mailAddress: string, signal: AbortSignal): Promise<U2fRegisteredDevice> {
-		const challenge = this._getChallenge()
-
-		const publicKeyCredentialCreationOptions: PublicKeyCredentialCreationOptions = {
-			challenge,
-			rp: {
-				name: "Tutanota",
-				id: this.rpId,
-			},
-			user: {
-				id: stringToUtf8Uint8Array(userId),
-				name: `${userId} ${mailAddress} ${name}`,
-				displayName: name,
-			},
-			pubKeyCredParams: [
-				{
-					alg: COSEAlgorithmIdentifierNames.ES256,
-					type: "public-key",
-				},
-			],
-			authenticatorSelection: {
-				authenticatorAttachment: "cross-platform",
-			},
-			timeout: WEBAUTHN_TIMEOUT_MS,
-			attestation: "none",
+	async canAttemptChallenge(challenge: U2fChallenge): Promise<{canAttempt: Array<U2fKey>, cannotAttempt: Array<U2fKey>}> {
+		// Whitelabel keys can ge registered other (whitelabel) domains.
+		// If it's a new Webauthn key it will match rpId, otherwise it will match legacy appId.
+
+		// Partition in keys that might work and which certainly cannot work.
```

Comment:
i'm not against putting this into a new function and calling it `partitionAsync`, instead of having an almost duplicate implementation hidden somewhere else with a comment explaining why it's necessary